# Lesson 1.6: Introduction to NumPy

## Introduction
**NumPy**, short for Numerical Python, is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.

Many computational and data science packages use NumPy as the main building block. It is a fundamental library for scientific computing in Python.

### Key Features of NumPy:
* **ndarray**: An efficient multidimensional array providing fast array-oriented arithmetic operations and flexible broadcasting capabilities.
* **Vectorization**: Mathematical functions for fast operations on entire arrays of data without having to write loops.
* **Linear Algebra**: Tools for random number generation, Fourier transforms, and matrix manipulation.
* **C API**: For connecting NumPy with libraries written in C, C++, or FORTRAN.

### Advantages over Python Lists:
1. **Contiguous Memory**: NumPy internally stores data in a contiguous block of memory, independent of other built-in Python objects. This allows for significantly faster access and manipulation.
2. **Vectorized Operations**: NumPy algorithms written in C can operate on this memory without type checking or other Python overhead, performing complex computations without slow `for` loops.

![numpy_vs_list](../assets/numpy_vs_python_list.png)

## Part 1: Performance Benchmark
To give you an idea of the performance difference, consider a NumPy array of one million integers and an equivalent Python list. We use the `%timeit` magic command to measure execution time.

In [1]:
# Can use the below to check for versions. Not required for ipynb cause the top right would have shown the versions.

import sys
print("Python version: ", sys.version)
print("Python location: ", sys.executable)

Python version:  3.10.20 (main, Mar 11 2026, 17:43:48) [Clang 20.1.8 ]
Python location:  /Users/exclusive/miniconda3/envs/pds/bin/python


In [2]:
import numpy as np

# np.arange() is a function in the NumPy library that creates an array of evenly spaced values within a specified range. It takes three parameters: start, stop, and step. The start parameter is the value at which the sequence begins (inclusive), the stop parameter is the value at which the sequence ends (exclusive), and the step parameter is the increment between each value in the sequence. For example, np.arange(0, 10, 2) would create an array of [0, 2, 4, 6, 8].
# we can use "_" to split the thousands for better readability. So 1_000_000 is the same as 1000000.
# Hover the funcations to see the "docs" for more info. Can also do help(np.arange) to see the docs in the console.
my_arr = np.arange(1_000_000)

# Creates a python list with values from 0 to 999,999.
my_list = list(range(1_000_000))

print("NumPy Vectorized Multiplication (my_arr * 2):")
%timeit my_arr2 = my_arr * 2

print("\nPython List Comprehension ([x * 2 for x in my_list]):")
%timeit my_list2 = [x * 2 for x in my_list]

NumPy Vectorized Multiplication (my_arr * 2):
1 ms ± 15.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Python List Comprehension ([x * 2 for x in my_list]):
37.9 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


NumPy Vectorized Multiplication (my_arr * 2):
1 ms ± 15.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Python List Comprehension ([x * 2 for x in my_list]):
37.9 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)



**Interpretation of results:**
- NumPy operation took 1 millisec while Python list took 37.9 millisec (~38x faster!)

## Part 2: The ndarray (N-dimensional array)
The `ndarray` is a fast, flexible container for large datasets. It is a multidimensional array of fixed size with **homogeneous** elements (all elements must be of the same type).

Every array has:
* **shape**: A tuple indicating the size of each dimension.
* **dtype**: An object describing the data type of the array.
* **ndim**: The number of dimensions (axes).

### ndarray illustration
![ndarray](../assets/numpy_ndarray.png)

🤔 (07-May-2026 Notes): Color and video images are usually represented as 3D arrays

In [7]:
# [DEMO] Creating arrays from sequences
data1 = [6, 7.5, 8, 0, 1]
arr1 = np.array(data1)

print(f"Array 1:\n{arr1}")
print(f"Shape: {arr1.shape}, Dtype: {arr1.dtype}, Dimensions: {arr1.ndim}")
print("\n")

data2 = [[1, 2, 3, 4], [5, 6, 7, 8]]
arr2 = np.array(data2)

print(f"Array 2:\n{arr2}")
print(f"Shape: {arr2.shape}, Dtype: {arr2.dtype}, Dimensions: {arr2.ndim}")

Array 1:
[6.  7.5 8.  0.  1. ]
Shape: (5,), Dtype: float64, Dimensions: 1


Array 2:
[[1 2 3 4]
 [5 6 7 8]]
Shape: (2, 4), Dtype: int64, Dimensions: 2


In [8]:
# 🤔 (07-May-2026 Example):
# Jagged arrays and rectagular requirements
jagged_data = [
    [1, 2, 3],
    [4, 5]
]

# The below code will fail
# jagged_arr = np.array(jagged_data)

# To create a jagged array, we can specify the dtype as 'object' to allow for arrays of different lengths.
jagged_arr_object = np.array(jagged_data, dtype=object)
print(f"Jagged Array (dtype=object):\n{jagged_arr_object}")
print(f"Shape: {jagged_arr_object.shape}, Dtype: {jagged_arr_object.dtype}, Dimensions: {jagged_arr_object.ndim}")

Jagged Array (dtype=object):
[list([1, 2, 3]) list([4, 5])]
Shape: (2,), Dtype: object, Dimensions: 1


### Data Types and Casting
NumPy supports specific numerical types like `int32`, `float64`, etc. You can explicitly convert an array from one `dtype` to another using the `astype` method.

**Note:** If you cast floating-point numbers to an integer `dtype`, the decimal part will be truncated.

In [9]:
# [DEMO] Casting arrays
# Maybe by default, Python assigns float64 to the array, but we can cast it to int32 using astype() method.
arr = np.array([3.7, -1.2, 0.5, 12.9])
print("Original:", arr)
print("Casted to int32:", arr.astype(np.int32))

# 🤔 (07-May-2026 Notes): Integer just drops the decimal part, aka round down

Original: [ 3.7 -1.2  0.5 12.9]
Casted to int32: [ 3 -1  0 12]


### 🛠️ Exercise 1: Creation & Casting
1. Create a 3x4 array of all ones using `np.ones()`.
2. Cast this array to `float32`.
3. Create an array of strings representing numbers: `['1.25', '-9.6', '42']`. Cast it to `float`.

In [11]:
# Your code here
# Create a 3x4 array of ones using np.ones()
arr = np.ones((3, 4), dtype=np.float64)
print("3x4 array of ones:\n", arr)

# Cast the above to float32
arr_float32 = arr.astype(np.float32)
print("\nCasted to float32:\n", arr_float32)
print("dtype before casting:", arr.dtype)
print("dtype after casting to float32:", arr_float32.dtype)

# Create an array of strings representing numbers: `['1.25', '-9.6', '42']`. Cast it to `float`.
arr_str_to_float = np.array(['1.25', '-9.6', '42']).astype(float)
print("\nArray of strings casted to float:\n", arr_str_to_float)
print("dtype of arr_str_to_float:", arr_str_to_float.dtype)

3x4 array of ones:
 [[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]

Casted to float32:
 [[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]
dtype before casting: float64
dtype after casting to float32: float32

Array of strings casted to float:
 [ 1.25 -9.6  42.  ]
dtype of arr_str_to_float: float64


In [13]:
# Common mistakes
one_array_mistake = np.ones(3,4) # This will raise an error because np.ones() expects a single argument for the shape, which should be a tuple. The correct way to create a 3x4 array of ones is np.ones((3, 4)).

str_array = np. array(['1.25', '-9.6', '42', 'string']) # This will raise an error because 'string' cannot be converted to a float. When trying to cast this array to float, it will fail due to the presence of the non-numeric string.
float_array_mistake = str_array.astype(float) # This will raise an error because 'string' cannot be converted to a float. The presence of the non-numeric string in the array will cause the astype() method to fail when trying to cast the entire array to float.


TypeError: Cannot interpret '4' as a data type

In [14]:
# np.zeroes
arr_zeroes = np.zeros((2, 3), dtype=np.int32)
print("\n2x3 array of zeroes:\n", arr_zeroes)

# np.full
arr_full = np.full((2, 3), 7, dtype=np.float64)
print("\n2x3 array filled with 7:\n", arr_full)


2x3 array of zeroes:
 [[0 0 0]
 [0 0 0]]

2x3 array filled with 7:
 [[7. 7. 7.]
 [7. 7. 7.]]


In [15]:
# try to change value of element
arr_zeroes[0, 0] = 5
print("\nModified array of zeroes:\n", arr_zeroes)

# changing element to other types
arr_full[0, 0] = 3.14
print("\nModified array filled with 7 (changed one element to 3.14):\n", arr_full)
print("dtype of arr_full after modification:", arr_full.dtype)


Modified array of zeroes:
 [[5 0 0]
 [0 0 0]]

Modified array filled with 7 (changed one element to 3.14):
 [[3.14 7.   7.  ]
 [7.   7.   7.  ]]
dtype of arr_full after modification: float64


In [ ]:
# because value is a float, the below will fail - incompatible type
arr_full[0,1] = "hello"

In [16]:
# Performance difference between datatypes
arr_64 = np.arange(1_000_000, dtype=np.float64)
arr_32 = np.arange(1_000_000, dtype=np.float32)

print(f"Performance of float64 array multiplication:")
%timeit arr_64 * 2

print(f"Performance of float32 array multiplication:")
%timeit arr_32 * 2

# show memory usage difference between float64 and float32
print(f"Memory usage of float64 array: {arr_64.nbytes} bytes")
print(f"Memory usage of float32 array: {arr_32.nbytes} bytes")

# in MB:
print(f"Memory usage of float64 array: {arr_64.nbytes / (1024 * 1024):.2f} MB")
print(f"Memory usage of float32 array: {arr_32.nbytes / (1024 * 1024):.2f} MB")


Performance of float64 array multiplication:
878 μs ± 32.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Performance of float32 array multiplication:
440 μs ± 123 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Memory usage of float64 array: 8000000 bytes
Memory usage of float32 array: 4000000 bytes
Memory usage of float64 array: 7.63 MB
Memory usage of float32 array: 3.81 MB


## Part 3: Arithmetic & Broadcasting
Arithmetic operations are applied as batch operations without for loops. **Broadcasting** describes how arithmetic works between arrays of different shapes.

![vectorization](../assets/vectorization.png)

Example: A scalar value being replicated (broadcast) to match the shape of a larger array.

In [17]:
# [DEMO] Arithmetic & Broadcasting
arr = np.array([[1., 2., 3.], [4., 5., 6.]])
print("Element-wise multiplication (arr * arr):\n", arr * arr)

# Broadcasting scalar multiplication, where the scalar is multiplied with each element of the array. 
# How this works is that the scalar value is "broadcast" across the entire array, allowing for element-wise operations without needing to explicitly create an array of the same shape as the original. This is a powerful feature of NumPy that enables efficient computations.
print("\nBroadcasting scalar (1 / arr):\n", 1 / arr)

Element-wise multiplication (arr * arr):
 [[ 1.  4.  9.]
 [16. 25. 36.]]

Broadcasting scalar (1 / arr):
 [[1.         0.5        0.33333333]
 [0.25       0.2        0.16666667]]


In [18]:
# More examples on broadcasting
# arr is a 2D, while 10 is 1D. When we do arr + 10, the scalar 10 is broadcasted to each element of the array, resulting in a new array where 10 has been added to every element of arr.
print("Adding scalar 10 to array (arr + 10):\n", arr + 10)

Adding scalar 10 to array (arr + 10):
 [[11. 12. 13.]
 [14. 15. 16.]]


## Part 4: Indexing and Slicing
One-dimensional arrays act similarly to Python lists. In 2D arrays, indexing can be done with `[row, column]` syntax. 

### 2D Array Indexing Syntax
![2d_array_indexing](../assets/ndarray_axis_index.png)

**Important:** Array slices are **views** on the original array. This means data is not copied, and modifications to the slice will be reflected in the source array.

In [20]:
# [DEMO] Slicing views (1D slicing)
arr = np.arange(10)
print("Original array:", arr)

# [start:stop:step]. Default will be used if unspecified. The default start is 0, the default stop is the length of the array, and the default step is 1.
arr_slice = arr[5:8]
print("Sliced array (arr[5:8]):", arr_slice)
arr_slice[1] = 12345                # Modifying the slice will modify the original array because slicing creates a view, not a copy. 
print("Original array modified via slice:", arr)

# Different from python list slicing which creates a new list. In NumPy, slicing returns a view of the original array, meaning that any modifications made to the slice will affect the original array since they share the same data in memory.
# we can use .copy() to create a copy of the sliced array, which will not affect the original array when modified.
arr_slice_copy = arr[5:8].copy()
arr_slice_copy[1] = 54321
print("Original array after modifying the copy of the slice:", arr)

Original array: [0 1 2 3 4 5 6 7 8 9]
Sliced array (arr[5:8]): [5 6 7]
Original array modified via slice: [    0     1     2     3     4     5 12345     7     8     9]
Original array after modifying the copy of the slice: [    0     1     2     3     4     5 12345     7     8     9]

First two rows, columns 1 onwards:
 [[2 3]
 [5 6]]


In [26]:
# [DEMO] 2D Slicing
arr2d = np.array([
    [1, 2, 3], 
    [4, 5, 6], 
    [7, 8, 9]])
print("\nFirst two rows, columns 1 onwards:\n", arr2d[:2, 1:])

# arr2d[row_slice, col_slice]
# start:stop:step for rows and columns. Default values will be used if unspecified. The default start is 0, the default stop is the length of the dimension, and the default step is 1.
# :2 means the first two rows (0 and 1), and 
# 1: means columns from index 1 to the end (columns 1 and 2). So the result is a 2D array containing the elements from the first two rows and the last two columns of arr2d.


First two rows, columns 1 onwards:
 [[2 3]
 [5 6]]


### 🛠️ Exercise 2: The Logic of Slicing
1. Select the first column of `arr2d` using a slice.
2. Set all values in the second row to 0.
3. **Socratic Prompt:** How does `arr2d[1]` differ from `arr2d[1, :]`? (Hint: check shapes)

In [28]:
# Your code here
# 1. Select the first column of `arr2d` using a slice.
print("1st column of arr2d:\n", arr2d[:, 0])
print("1st column of arr2d:\n", arr2d[:, :1])
# For the entire column of rows, we set ":" for all rows, and "0" for the 1st column, index = 0.

# 2. Set all values in the second row to 0.
arr2d[1, :] = 0
print("\narr2d after setting second row to 0:\n", arr2d)
# For the second row, we set "1" for the row index, and ":" for all columns to set all values in that row to 0.

# 3. **Socratic Prompt:** How does `arr2d[1]` differ from `arr2d[1, :]`? (Hint: check shapes)
# The below selects the 2nd element of the array. Since arr2d is a 2D array, arr2d[1] will return the entire second row as a 1D array. The shape of arr2d[1] will be (3,), indicating that it is a 1D array with 3 elements. The dimensions (ndim) of arr2d[1] will be 1, since it is a 1D array.
print("arr2d[1]:", arr2d[1])
print("Shape of arr2d[1]:", arr2d[1].shape)
print("Dimensions of arr2d[1]:", arr2d[1].ndim)

# On the other hand, arr2d[1, :] will also return the second row, but it will maintain the 2D structure of the original array. The shape of arr2d[1, :] will be (1, 3), indicating that it is a 2D array with 1 row and 3 columns. The dimensions (ndim) of arr2d[1, :] will be 2, since it is a 2D array.
print("arr2d[1, :]:", arr2d[1, :])
print("Shape of arr2d[1, :]:", arr2d[1, :].shape)
print("Dimensions of arr2d[1, :]:", arr2d[1, :].ndim)

print("arr2d[1:2, :]:", arr2d[1:2, :])
print("Shape of arr2d[1:2, :]:", arr2d[1:2, :].shape)
print("Dimensions of arr2d[1:2, :]:", arr2d[1:2, :].ndim)

1st column of arr2d:
 [1 0 7]
1st column of arr2d:
 [[1]
 [0]
 [7]]

arr2d after setting second row to 0:
 [[1 2 3]
 [0 0 0]
 [7 8 9]]
arr2d[1]: [0 0 0]
Shape of arr2d[1]: (3,)
Dimensions of arr2d[1]: 1
arr2d[1, :]: [0 0 0]
Shape of arr2d[1, :]: (3,)
Dimensions of arr2d[1, :]: 1
arr2d[1:2, :]: [[0 0 0]]
Shape of arr2d[1:2, :]: (1, 3)
Dimensions of arr2d[1:2, :]: 2


## Part 5: Boolean Indexing
Like arithmetic operations, comparisons (such as `==`) with arrays are vectorized. This yields a boolean array which can be used to filter data.

In [29]:
# [DEMO] Filtering scores
names = np.array(['Bob', 'Joe', 'Will', 'Bob', 'Will', 'Joe', 'Joe'])
scores = np.array([[75, 80], [85, 90], [95, 100], [100, 77], [85, 92], [95, 80], [72, 80]])

# A mask is a boolean array indicating which elements to select. List of True False values in the array.
bob_mask = (names == 'Bob')
print("Mask:", bob_mask)
print("Bob's scores:\n", scores[bob_mask])

Mask: [ True False False  True False False False]
Bob's scores:
 [[ 75  80]
 [100  77]]


### 🛠️ Exercise 3: Complex Filtering
1. Select all scores where the name is NOT 'Bob'.
2. Select scores for 'Bob' or 'Will' using the `|` operator.
3. Find all scores less than 80 and set them to 0.

In [36]:
# Your code here
# 1. Select all scores where the name is NOT 'Bob'.
print("Score not from Bob: ", names != 'Bob')

# 2. Select scores from either 'Bob' or 'Will' using the | operator.
mask = (names == 'Bob') | (names == 'Will')
print("Scores from either Bob or Will: ", scores[mask])

# 3. Find all scores less than 80 and set them to 0.
# Practice copy concept just to practice this topic
score_copy = scores.copy()
scores_below_80 = score_copy < 80           # Setting boolean to the array
score_copy[scores_below_80] = 0             # Based on the boolean array, set the corresponding elements in score_copy to 0
print("\nScores with values less than 80 set to 0:\n", score_copy)


Score not from Bob:  [False  True  True False  True  True  True]
Scores from either Bob or Will:  [[ 75  80]
 [ 95 100]
 [100  77]
 [ 85  92]]

Scores with values less than 80 set to 0:
 [[  0  80]
 [ 85  90]
 [ 95 100]
 [100   0]
 [ 85  92]
 [ 95  80]
 [  0  80]]


## Part 6: Universal Functions (ufuncs) and Methods
A **ufunc** is a function that performs element-wise operations on data in ndarrays. 

* **Unary ufuncs**: Take one array (e.g., `sqrt`, `exp`).
* **Binary ufuncs**: Take two arrays (e.g., `add`, `maximum`).
* **Statistical Methods**: `mean`, `sum`, `std` can be computed over the entire array or along an axis.

In [37]:
# [DEMO] Statistical Methods
# Create a random 3x4 array using standard normal distribution.
arr = np.random.randn(3, 4)
print("Random Array:\n", arr)
print("\nMean down rows (axis=0):", arr.mean(axis=0))               # axis 0 = vertical; 1 horitzontal direction
print("Sum across columns (axis=1):", arr.sum(axis=1))

Random Array:
 [[ 2.0292231   0.5732871   0.42692091  0.58566038]
 [-1.00024286 -0.47756636 -2.10629587 -0.85707447]
 [ 1.26965397 -0.56985036  1.50066281  0.93364311]]

Mean down rows (axis=0): [ 0.7662114  -0.15804321 -0.05957072  0.22074301]
Sum across columns (axis=1): [ 3.61509149 -4.44117956  3.13410953]


## Part 7: Linear Algebra
Linear algebra operations, like matrix multiplication, are crucial for many data science algorithms. Multiplying two arrays with `*` is an element-wise product; for matrix multiplication, use `.dot()` or the `@` operator.

![matrix_multiplication](../assets/matrix_multiplication.png)

In [32]:
# [DEMO] Matrix Multiplication
x = np.array([[1, 2, 3], [4, 5, 6]])
y = np.array([[6, 23], [-1, 7], [8, 9]])

print("Matrix product (x @ y):\n", x @ y)

Matrix product (x @ y):
 [[ 28  64]
 [ 67 181]]


### 🛠️ Exercise 4: Reshaping & Statistics
1. Create an array of 15 integers using `arange(15)` and reshape it to `(3, 5)`.
2. Calculate the average value of each row.
3. Use `np.unique()` to find distinct elements in an array of your choice.
4. Transpose the reshaped array using `.T` and check the new shape.

In [41]:
# Your code here
# 1. Create an array of 15 integers using `arange(15)` and reshape it to `(3, 5)`.
arr_of_15 = np.arange(15)
print("Array of 15 elements:", arr_of_15)
print("Shape of arr_of_15:", arr_of_15.shape)

# 2. Calculate the average value of each row.

reshaped_arr = arr_of_15.reshape(3, 5)
print("Reshaped array to 3x5:\n", reshaped_arr)
print("Shape of reshaped_arr:", reshaped_arr.shape)
# Finding averge clue of each ros.
arr_row_avg = arr_of_15.mean(axis=0)


# 3. Use `np.unique()` to find distinct elements in an array of your choice.
# Finding distinct elements in array of choice.
arr_unique = np.unique(reshaped_arr)
print("Array of unique elements:\n", arr_unique)



# 4. Transpose the reshaped array using `.T` and check the new shape.# 
arr_transposed = reshaped_arr.T
print("Transposed array:\n", arr_transposed)
print("Shape of transposed array:", arr_transposed.shape)


Array of 15 elements: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
Shape of arr_of_15: (15,)
Reshaped array to 3x5:
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]
Shape of reshaped_arr: (3, 5)
Array of unique elements:
 [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
Transposed array:
 [[ 0  5 10]
 [ 1  6 11]
 [ 2  7 12]
 [ 3  8 13]
 [ 4  9 14]]
Shape of transposed array: (5, 3)
